### PR TITLE
GH-116939: Rewrite binarysort()

### PR DESCRIPTION
**Pull request by tim-one**
* Issue: gh-116939

### CODE REVIEW

**Comment by tim-one** [comment]
To my eyes, this addresses the annoyances listed in the linked issue. Under MSVC I see no real difference in speed, although probably by sheer luck most runs were slightly faster after the change.

File: Objects/listobject.c [hunk 1]
```diff
@@ -1665,53 +1674,53 @@ struct s_MergeState {
...
-            p = start + offset;
-            pivot = *p;
-            l += offset;
-            for ( ; p > l; --p)
-                *p = *(p-1);
-            *l = pivot;
+                L = M + 1;
+        } while (L < R);
+        assert(L == R);
+        /* a[:L] holds all elements from a[:ok[ <= pivot now, so pivot belongs
```
Comment (pochmann3) [inline Objects/listobject.c]: By now you've used `)`, `]` and `[` on the right side, but I think in all three cases (`s[:ok)` and these two), the right index is exclusive, right?

Comment (tim-one) [reply Objects/listobject.c]: Yes, this is meant to be Python slice notation.. Alas, I'm typing with inadequate light on a new keyboard, and these keys seem endlessly confusable ☹️ .

**Comment by tim-one** [comment]
Note recent commit:

```none
Add preprocesor cruft so that you can very easily try either
binary or regular insertion sort. And for binary insertion
sort, switch to a different approach that tries to eliminate the
inner branch by generating conditional move instructions
instead. Alas, Visual Studio will not show me the machine
code generated in a Release build - the compiler and/or linker
apparently optimized it so much it can't relate it to the
source code :-(
```

**Comment by tim-one** [comment]
Anyone have an idea about why "Tests / CIFuzz (memory) (pull_request) " could fail? I don't know anything about it. Clicking on "Details" brings up a giant amount of info. It appears to be failing when building (not running!) "a fuzzer", and about 2000 thousand lines down, near the end of that section:

```none
./Programs/_freeze_module getpath ./Modules/getpath.py Python/frozen_modules/getpath.h
Segmentation fault (core dumped)
make: *** [Makefile:1650: Python/frozen_modules/getpath.h] Error 139
```

I'd ignore it, except Visual Studio tries building some freeze-related thing as a matter of course when building CPython, and it tries to do some sorting (but seemingly with a 32-bit(!) CPython), and errors while developing _this_ code often died during that step.

File: Objects/listobject.c [hunk 2]
```diff
@@ -1683,7 +1683,7 @@ binarysort(MergeState *ms, const sortslice *ss, Py_ssize_t n, Py_ssize_t ok)
     Py_ssize_t L, M, R;
 
     assert(0 <= ok && ok <= n && 1 <= n && n <= MAX_MINRUN);
-    /* assert s[:ok) is sorted */
+    /* assert s[:ok] is sorted */
```
Comment (pochmann3) [inline Objects/listobject.c]: Is `s` a typo? I'd use `a` throughout. That is, `a[:ok]` here and also in the later invariants replace `0:L` with `a[0:L]` etc.

Comment (tim-one) [reply Objects/listobject.c]: Thanks. It's a typo.

**Comment by hugovk** [comment]
CIFuzz hasa started failing for other PRs too: https://github.com/python/cpython/issues/116886.

**Pull request merged**